### ISSUE TITLE
iOS 10 > Display an image in notification

### ISSUE THREAD

**Issue by appzer** [phonegap/phonegap-plugin-push#1347]
Hello,

it is possible to display an image in notifications on iOS >10,0 with this plugin?

Kevin

**Comment by macdonst** [comment]
@appzer not currently but PR's for this feature would be welcome.

**Comment by distante** [comment]
@macdonst this does not just depends on the push payload?

**Comment by appzer** [comment]
right, you also have to add a notification extension to your project in xcode

**Comment by macdonst** [comment]
OMG, this is such a PITA. I've been prototyping some stuff and it won't work on PhoneGap Build until cordova-ios can support multiple targets but there is a future in this enhancement.

**Comment by pavanst77** [comment]
is Cordova-ios suporting now?

**Comment by jcesarmobile** [comment]
no

**Comment by npostulart** [comment]
Any way we can get this feature running in the near future. Highly needed the support for images in iOS notifications.

**Comment by pliablepixels** [comment]
Great warriors of cordova-land, any thoughts on who I need to follow up to get picture notifications working in iOS? I don't have the native skills to do a PR. @macdonst @jcesarmobile

**Comment by ChrisEelmaa** [comment]
Maybe it will help someone, I got proof of concept with 30  minutes (how to display images), here's what I did:

I went to XCode, 

`File->New Target-> Notification Service Extension`,

copy/pasted the following code over the existing code (`NotificationService.swift`), this code basically knows how to turn custom APN image key in to actual notification attachment:

```
import UserNotifications

class NotificationService: UNNotificationServiceExtension {
    
    var contentHandler: ((UNNotificationContent) -> Void)?
    var bestAttemptContent: UNMutableNotificationContent?
    
    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
        self.contentHandler = contentHandler
        self.bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
        guard let content = bestAttemptContent else {
            #if DEBUG
            exitGracefully("bestAttemptContent not a UNMutableNotificationContent")
            #endif
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            let userInfo: [AnyHashable: Any] = request.content.userInfo
            content.attachments = self?.attachmentsFor(userInfo) ?? []
            guard let copy = self?.bestAttemptContent else {
                #if DEBUG
                self?.exitGracefully("bestAttemptContent is nil")
                #endif
                return
            }
            contentHandler(copy)
        }
    }
    
    override func serviceExtensionTimeWillExpire() {
        if let contentHandler = contentHandler, let content = self.bestAttemptContent {
            contentHandler(content)
        }
    }
    
    /// Helper function to extract Attachments from an userInfo object
    ///
    /// - Attribute userInfo: The user info extracted from the notification,
    ///                       which should contain one of the available keys to create
    ///                       a rich push notification.
    private func attachmentsFor(_ userInfo: [AnyHashable: Any]) -> [UNNotificationAttachment] {
        if let attachmentURLString = userInfo["image-url-png"] as? String {
            guard let attachmentURL = URL(string: attachmentURLString),
                let imageData = try? Data(contentsOf: attachmentURL),
                let attachment = self.save("image.png", data: imageData, options: nil) else {
                    #if DEBUG
                    self.exitGracefully("PNG was not saved properly")
                    #endif
                    return []
            }
            return [attachment]
        } else if let attachmentURLString = userInfo["image-url-jpg"] as? String {
            guard let attachmentURL = URL(string: attachmentURLString),
                let imageData = try? Data(contentsOf: attachmentURL),
                let attachment = self.save("image.jpg", data: imageData, options: nil) else {
                    #if DEBUG
                    self.exitGracefully("JPG was not saved properly")
                    #endif
                    return []
            }
            return [attachment]
        } else if let attachmentURLString = userInfo["image-url-gif"] as? String {
            guard let attachmentURL = URL(string: attachmentURLString),
                let imageData = try? Data(contentsOf: attachmentURL),
                let attachment = self.save("image.gif", data: imageData, options: nil) else {
                    #if DEBUG
                    self.exitGracefully("GIF was not saved properly")
                    #endif
                    return []
            }
            return [attachment]
        }
        
        return []
    }
    
    /// Save data object onto disk and return an optional attachment linked to this path.
    ///
    /// - Attributes:
    ///   - identifier: The unique identifier of the attachment.
    ///                 Use this string to identify the attachment later. If you specify an
    ///                 empty string, this method creates a unique identifier string for you.
    ///   - data: The data stored onto disk.
    ///   - options: A dictionary of options related to the attached file.
    ///              Use the options to specify meta information about the attachment,
    ///              such as the clipping rectangle to use for the resulting thumbnail.
    private func save(_ identifier: String, data: Data, options: [AnyHashable: Any]?) -> UNNotificationAttachment? {
        // Create paths
        let directory = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(ProcessInfo.processInfo.globallyUniqueString, isDirectory: true)
        let fileURL = directory.appendingPathComponent(identifier)
        // Write data on disk
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)
        try? data.write(to: fileURL, options: [])
        // Create Notification attachment
        return try? UNNotificationAttachment(identifier: identifier, url: fileURL, options: options)
    }
    
    /// Something went wrong, so maybe we want to clean up and present a fallback to the user.
    ///
    /// - Attribute reason: The reason why something went wrong.
    private func exitGracefully(_ reason: String = "") {
        guard let copy = bestAttemptContent?.mutableCopy() as? UNMutableNotificationContent else { return }
        copy.title = reason
        contentHandler?(copy)
    }
    
}
```

That's all I did, now I can send push notification with image included, as such:

```
{
	aps: {
		alert: {
			title : "Hooraayy!",
			body: `${myProfile.get('name')} has sent you a new message`
		},
	   "mutable-content" : 1,
	}

	'image-url-png':  photos[0].url
 }
```

**Comment by pliablepixels** [comment]
@ChrisEelmaa thank you!
I managed to get the extension compiled and running. That being said, I use FCM with iOS and I don't see any images on my iOS device.


Here is what my server is sending. Do you see any issue?

 Final JSON being sent is: 
```

{
    "to":"cFBL1a1p3qw:APA91b<deleted>Gi",
     "notification":{"title":"Unfinished Alarm (306136)",
                               "body":"Forced Web at 02:23 PM, 27-Dec",
                               "badge":"1",
                               "sound":"default"},
        "data":  {
         "mediaUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/2/2a/FloorGoban.JPG/1024px-FloorGoban.JPG",
 "summaryText":"306136"},
     "content_available":true,
     "mutable_content":true}
```

**Comment by halie412** [comment]
Hello, i follow the two steps dicted by @ChrisEelmaa 

1- I add an new target (Notification Service Extension)
2- edit file NotificationService.swift
additionally i reated all nesesary certs and configs to avoid errors in Xcode when try to Archive or run the project

I have my app correctly running in an iPad 3 with iOS 10.3.3

But when i send an notification with the content:

{"aps":{"alert":{"title":"un titulo","body":"un mensaje"},"mutable-content":1},"image-url-png":"https://myweb/my-image .png"}

I receibe the notification but i can`t see the image

**Comment by stingray91** [comment]
@ChrisEelmaa @pliablepixels I would love to give your solution a try, but I'm facing problems with creating a new notification service extension.. What should I put in the product name section?
If I were to give it a new name, I would face info.plist missing during "ionic cordova build ios --release --buildFlag="-UseModernBuildSystem=0"" in CLI.

Thanks!

<img width="714" alt="image" src="https://user-images.githubusercontent.com/12985374/53614047-67800880-3c13-11e9-90fc-4016f8c89c71.png">

**Comment by pliablepixels** [comment]
I faced two issues (but not your exact issue):

a) Debug builds would always work, but production builds would fail with `The bundle at … contains disallowed file 'Frameworks'`
b) Command line building broke (works perfectly fine without the extension)

As far as a) is concerned, there are multiple approaches to address it, one of which is to write a run script to remove the Frameworks folder and the other was to include swift libraries in the container but not in the extension. These are approaches from Stack Overflow (you can google for it). I opted for an easier way, I replaced the swift code with an [Objective C code](https://github.com/pliablepixels/zmNinja/blob/master/www/external/NotificationService.m) (I am using `Legacy Build` due to [another Cordova-iOS 4.x issue](https://github.com/apache/cordova-ios/issues/412) that has issues including profiles. This has apparently been addressed in 5) that got rid of this issue completely

b) My guess is given Cordova-iOS build process doesn't support multiple targets as of today (see comments earlier) I'm not sure command line build will work at all (and if it does, let me know what your build file looks like)

**Comment by stingray91** [comment]
@pliablepixels  I see, I shall compile in production mode the first thing next week and see how it goes. Thanks!

**Comment by ChrisEelmaa** [comment]
@stingray91 try removing/adding the ios platform project again. 

I can confirm that for me "cordova build ios" works without problems (note that I've enabled Legacy build in xcode settings, so no need for any arguments). I didn't do anything special for it. Seems like your build is somehow messed up. As said, start from zero: remove ios project and 'cordova platform add ios' again.

That said, I've ditched cordova/ionic in favour of react-native and I am very pleased. good luck.

**Comment by guswls1846** [comment]
good!
Chris Elmama's method is right. 
After create Notification Service Extension Copy and paste ChrisEelmaa's code!
If userInfo["image-url-jpg"]  or userInfo["mediaurl"] does not work like halie412  Please note that info.plist of  Notification Service Extension  add AppTransport Security Settings and nuder  Allow Arbitrary Loads = YES

![image](https://user-images.githubusercontent.com/32049764/55778324-94aabb00-5add-11e9-9131-92d27b194fbe.png)

My PHP payload
  $fields = array();
	            $fields['notification']['title'] = (!empty($message));
	            $fields['notification']['body'] = $content;
	            $fields['notification']['sound'] = 'default';
	            $fields['notification']['badge'] = '0';
	            $fields['content-available'] = true;
	            $fields['priority'] = 'high';
                    $fields['data']['image-url-jpg'] = $picture;
                    $fields['notification']['mutable_content'] = true;

**Comment by MatejZmitko** [comment]
For those using AWS SNS. 

1. Create a new target (Notification Service Extension)
2. Copy and replace the generated swift code with the code from @ChrisEelmaa
3. Make sure to add (click on a + button at the `Information Property List level`) `App Transport Security Settings` and `Allow Arbitrary Loads` inside and change the Boolean to YES exactly like @guswls1846 
4. go to your AWS SNS console and send this (or similar) chunk of payload:
```
{
  "APNS_SANDBOX": "{\"aps\":{\"alert\":{\"title\":\"A short string describing the purpose of the notification\",\"body\":\"The text of the alert message\"},\"mutable-content\":\"1\"},\"image-url-jpg\":\"https://i.imgur.com/t4WGJQx.jpg\"}"
}

```

**Comment by 2xSamurai** [comment]
> I faced two issues (but not your exact issue):
> 
> a) Debug builds would always work, but production builds would fail with `The bundle at … contains disallowed file 'Frameworks'`
> b) Command line building broke (works perfectly fine without the extension)
> 
> As far as a) is concerned, there are multiple approaches to address it, one of which is to write a run script to remove the Frameworks folder and the other was to include swift libraries in the container but not in the extension. These are approaches from Stack Overflow (you can google for it). I opted for an easier way, I replaced the swift code with an [Objective C code](https://github.com/pliablepixels/zmNinja/blob/master/www/external/NotificationService.m) (I am using `Legacy Build` due to [another Cordova-iOS 4.x issue](https://github.com/apache/cordova-ios/issues/412) that has issues including profiles. This has apparently been addressed in 5) that got rid of this issue completely
> 
> b) My guess is given Cordova-iOS build process doesn't support multiple targets as of today (see comments earlier) I'm not sure command line build will work at all (and if it does, let me know what your build file looks like)

Your Objective C code link is not working. Can you please share it again. And for the command line part. Did you find any solution? I am also stuck there.

**Comment by pliablepixels** [comment]
Linked fixed in post ([this](https://github.com/pliablepixels/zmNinja/blob/master/etc/NotificationService.m))

I still manually change the bundle ID in Xcode.

**Comment by 2xSamurai** [comment]
> Linked fixed in post ([this](https://github.com/pliablepixels/zmNinja/blob/master/etc/NotificationService.m))
> 

Will changing from swift to Objective C help me solve the command line build fail issue. I am using Ionic v1 and also stuck with cordova ios 4.x

> I still manually change the bundle ID in Xcode.
why should i do this? Sorry for asking.

**Comment by 2xSamurai** [comment]
> Linked fixed in post ([this](https://github.com/pliablepixels/zmNinja/blob/master/etc/NotificationService.m))
> 
> I still manually change the bundle ID in Xcode.

Do I need to do something in 'NotificationService.h' file as well?

**Comment by VM8198** [comment]
I had done all Setting in xCode, but still image is not displayed in notification.

**Comment by MatejZmitko** [comment]
> I had done all Setting in xCode, but still image is not displayed in notification.

Thats a very basic information. Do you receive other notifications (eg. with plain text) so that you can be sure that your certificates work correctly? How do you send notifications to you device? What do you mean by 'all settings', did you create the notification extension entry point?

**Comment by VM8198** [comment]
hello @MatejZmitko Thanks for responding, as you mentioned i am getting all notifications sent from fcm and from other devices also but the notifications having images and text received but the image is not displayed,
i have configured everything properly in XCode and FCM but i am not getting where is the issue
please help........

**Comment by MatejZmitko** [comment]
> hello @MatejZmitko Thanks for responding, as you mentioned i am getting all notifications sent from fcm and from other devices also but the notifications having images and text received but the image is not displayed,
> i have configured everything properly in XCode and FCM but i am not getting where is the issue
> please help........

Did you check some of the similar issues (eg. [this](https://stackoverflow.com/questions/42531730/fcm-rich-push-notification-payload-for-ios) one)? Are you sure your FCM notification payload (eg. `"image-url-png":`) in FCM corresponds to the xCode declaration (eg. `image-url-png`)? Did you set your `mutable-content` to true/1?

**Comment by VM8198** [comment]
Yes, In my Notification payload mutable-content = 1  and image are also there.

**Comment by guswls1846** [comment]
@VM8198  hello
 If you set the Arbitrary Loads to YES like @guswls1846 did, you will enable exceptions for HTTP sources to load, please see the documentation that says:

**Comment by VM8198** [comment]
@guswls1846 everything is set up as you suggested and as others answers but I m not getting why the images are not working in notifications

**Comment by guswls1846** [comment]
1. Are you sure your target service is Notification Service Extension?  
Notification Service Extension and Notification Service Content are different
2. Did you embedded content activate it when adding a target?
![image](https://user-images.githubusercontent.com/32049764/68837260-8bcb3900-06ff-11ea-816e-c82d2118d75e.png)
3. your NotificationService.swift and swift file in class name is same? 
example NotificaionService.swift == class NotificationService
4.tell me your version of ios Notification Service Extension is ios version 10 or later
5. Can you show me your payload? 
Receiving an image url in the payload is almost successful. Just like your Notification Service Extension can't catch the payload's url
It is recommended to change the payload format.
https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md#fcm-payload-details

In my case,	
"notification": {
		"title": "Test title", /* The notification's title */
		"body": "Test body.", /* The notification's body text */
                "mutable_content":true
	                                           .
	                                           .
	                                           .
	},
"data": {
              "image-url-jpg":"https://~~~"
	},
"content-available": true,
"priority" :"high"

**Comment by VM8198** [comment]
@guswls1846 I had created notificationService Extension, and also embedded content activate.
This is my payload:
{“fcm_options”:{“image”:“https://qph.fs.quoracdn.net/main-qimg-2409a441a71fdc85cf987c349d364cb4.webp”},
“gcm.message_id”:“1573457473685383”,
“google.c.a.e”:“1”,
“wasTapped”:true,
“aps”:{“mutable-content”:“1”,
"content-available":"1",
“alert”:{“title”:“ffff”},
“category”:“FCM_PLUGIN_ACTIVITY”,
“sound”:“default”}}

**Comment by guswls1846** [comment]
@VM8198  your NotificationService.swift code the same as @ChrisEelmaa ?
my NotificastionService.swift code  exactly like a ChrisEelmaa
your payload form differs base on ChrisEelmaa code and paylod
example 
"image"-> "image-url-jpg"  and What is fcm_options??
“mutable-content”:“1”->“mutable-content”:1

**Comment by VM8198** [comment]
@guswls1846 In my payload fcm-options is generated by  FCM.

**Comment by souly1** [comment]
well, for me tested on device and works well but when trying to archive fails with "Swift Compiler Error":
'Cordova/CDV.h' file not found in Bridging-Header.h :/
